src/packet/tcp.rs

```diff
@@ -0,0 +1,197 @@
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! TCP packet abstraction
+
+use packet::Packet;
+use packet::HasPseudoheader;
+use packet::checksum::rfc1071_checksum;
+use pnet_macros::types::*;
+
+/// Represents a TCP Packet
+#[packet]
+pub struct Tcp {
+    source: u16be,
+    destination: u16be,
+    sequence: u32be,
+    acknowledgement: u32be,
+    data_offset_and_reserved: u8,
+    control_bits: u8,
+    window: u16be,
+    checksum: u16be,
+    urgent_pointer: u16be,
+    #[length_fn = "tcp_options_length"]
+    options: Vec<TcpOptions>,
+    #[length_fn = "tcp_padding_length"]
+    padding: Vec<TcpPadding>,
+    #[payload]
+    payload: Vec<u8>
+}
+
+fn tcp_options_length<'a>(tcp: &TcpPacket<'a>) -> usize {
+    /* The data offset field specifies the total size of the TCP header
+       in 32-bit words. The minimum size of a TCP header is 20 bytes and thus
+       TCP headers must have data offset set to 5 or more.
+     */
+    let mut v = tcp.get_data_offset_and_reserved();
+    v  = v >> 4; // get rid of reserved bits
+    v = v - 5; // remove the size of the minimal tcp header of 20 bytes
+    v = v * 4; // multiple the remaining word count by 4 to get the byte count
+    return v as usize;
+}
+
+/// Represents the TCP Option fields
+#[packet]
+pub struct TcpOptions {
+    kind: u8,
+    #[payload]
+    data: Vec<u8>
+}
+
+fn tcp_padding_length<'a>(tcp: &TcpPacket<'a>) -> usize {
+    /* The TCP header padding is used to ensure that the entire header
+       ends on a 32 bit boundary.
+     */
+    //return tcp.get_data_offset_and_reserved() as usize % 4;
+    return 0
+}
+
+/// Represents the TCP header padding
+#[packet]
+pub struct TcpPadding {
+    #[payload]
+    data: Vec<u8>
+}
+
+/// Calculates the checksum of a TCP packet
+/// The passed in TcpPacket must have it's initial checksum value set to zero.
+pub fn checksum<'a, T: HasPseudoheader>(packet: &TcpPacket<'a>, encapsulating_packet: T) -> u16be {
```

Comment:
Does `encapsulating_packet` really need to take ownership? How well does this work? I've previously avoided having multiple packets existing for the same buffer for ownership reasons - it can become a hassle trying to work around it.